File: 2300-SuccessfulPairsofSpellsandPotions/2300-SuccessfulPairsofSpellsandPotions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int n = spells.length;
        int m = potions.length;
        Arrays.sort(potions);
        int[] res = new int[n];
        int index = 0;
        for(int spell: spells){
            int left = 0, right = m;
            //finding the last potion so that spell * potion < success
            while(left < right){
                // System.out.println("left: " + left + "; right: " + right);
                int mid = left + (right - left) / 2;
                int potion = potions[mid];
                if((long)spell * (long)potion >= success){ // right pivot doesn't move when mid is valid
                    right = mid;
                } else {
                    left = mid + 1; //left will move if mid is invalid
                }
            }
            // System.out.println("left: " + left + "; right: " + right);
            res[index] = m - right;
            index++;
        }
        return res;
    }
}

File: 2300-SuccessfulPairsofSpellsandPotions/2300-SuccessfulPairsofSpellsandPotions.py
# Last updated: 5/10/2025, 11:17:52 AM
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        left, right = 0, len(citations) - 1 
        if citations[-1] == 0:
            return 0
        while left <= right:
            mid = (left + right) // 2
            if citations[mid] >= len(citations) - mid:
                right = mid - 1
            else:
                left = mid + 1
        print("left, right:",left,"; " ,right)
        return len(citations) - left
        